BiggerReactors/src/main/java/net/roguelogix/biggerreactors/client/jei/BiggerReactorsJEIPlugin.java

```diff
@@ -0,0 +1,77 @@
+import net.minecraft.util.ResourceLocation;
+import net.minecraftforge.fluids.FluidStack;
+import net.roguelogix.biggerreactors.BiggerReactors;
+import net.roguelogix.biggerreactors.classic.reactor.ReactorModeratorRegistry;
+import net.roguelogix.biggerreactors.classic.reactor.blocks.ReactorTerminal;
+import net.roguelogix.biggerreactors.classic.turbine.TurbineCoilRegistry;
+import net.roguelogix.biggerreactors.classic.turbine.blocks.TurbineTerminal;
+import net.roguelogix.biggerreactors.client.jei.classic.reactor.BlockModeratorCategory;
+import net.roguelogix.biggerreactors.client.jei.classic.reactor.FluidModeratorCategory;
+import net.roguelogix.biggerreactors.client.jei.classic.turbine.CoilCategory;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@JeiPlugin
+public class BiggerReactorsJEIPlugin implements IModPlugin {
+
+    @Nonnull
+    public ResourceLocation getPluginUid() {
+        return new ResourceLocation(BiggerReactors.modid, "jei_plugin");
+    }
+
+    @Override
+    public void registerCategories(IRecipeCategoryRegistration registration) {
+        IGuiHelper guiHelper = registration.getJeiHelpers().getGuiHelper();
+        registration.addRecipeCategories(new CoilCategory(guiHelper));
+        registration.addRecipeCategories(new BlockModeratorCategory(guiHelper));
+        registration.addRecipeCategories(new FluidModeratorCategory(guiHelper));
+    }
+
+    @Override
+    public void registerRecipeCatalysts(IRecipeCatalystRegistration registration) {
+        registration.addRecipeCatalyst(new ItemStack(TurbineTerminal.INSTANCE), CoilCategory.UID);
+        registration.addRecipeCatalyst(new ItemStack(ReactorTerminal.INSTANCE), BlockModeratorCategory.UID);
+        registration.addRecipeCatalyst(new ItemStack(ReactorTerminal.INSTANCE), FluidModeratorCategory.UID);
+    }
+
+    @Override
+    public void registerRecipes(IRecipeRegistration registration) {
+        List<CoilCategory.Recipe> recipes = TurbineCoilRegistry.getImmutableRegistry().entrySet().stream()
+                .map(e -> new CoilCategory.Recipe(new ItemStack(e.getKey().asItem()), e.getValue()))
+                .collect(Collectors.toList());
+
+        registration.addRecipes(recipes, CoilCategory.UID);
+
+
+        List<FluidModeratorCategory.Recipe> fluidModeratorRecipes = new ArrayList<>();
+        List<BlockModeratorCategory.Recipe> blockModeratorRecipes = new ArrayList<>();
+
+        ReactorModeratorRegistry.getImmutableRegistry().forEach((block, moderatorProperties) -> {
+            if (block instanceof FlowingFluidBlock) {
+                FlowingFluidBlock fluidBlock = (FlowingFluidBlock) block;
+                FluidStack stack = new FluidStack(fluidBlock.getFluid(), 1000);
+                fluidModeratorRecipes.add(new FluidModeratorCategory.Recipe(stack, moderatorProperties));
+            } else if (!(block instanceof AirBlock)) {
```

Comment:
Air is a valid moderator, and has its own properties, its shit, but it does have them

Reply:
Problem is JEI considers an ItemStack containing `minecraft:air` an empty/invalid recipe, and drops it. Not really sure how to work around that other than excluding them entirely.

Reply:
alll right then, guess there isn't much that can be done about that